windows/src/engine/kmcomapi/util/Keyman.System.BaseKeyboard.pas

```diff
@@ -0,0 +1,183 @@
+    $00010415, // Polish (214)
+    $00010416, // Portuguese (Brazilian ABNT2)
+    $00010418, // Romanian (Standard)
+    $0001041b, // Slovak (QWERTY)
+    $0001041f, // Turkish F
+    $00010426, // Latvian (QWERTY)
+    $00010427, // Lithuanian
+    $0001042c, // Azerbaijani (Standard)
+    $0001042e, // Sorbian Extended
+    $0001043a, // Maltese 48-Key
+    $0001043b, // Sami Extended Norway
+    $0001045d, // Inuktitut - Naqittaut
+    $0001080c, // Belgian (Comma)
+    $0001083b, // Finnish with Sami
+    $00011009, // Canadian Multilingual Standard
+    $00011809, // Gaelic
+    $00020405, // Czech Programmers
+    $00020409, // United States-International
+    $00020418, // Romanian (Programmers)
+    $00020426, // Latvian (Standard)
+    $00020427, // Lithuanian Standard
+    $0002042e, // Sorbian Standard
+    $0002083b, // Sami Extended Finland-Sweden
+    $00030408, // Greek (220) Latin
+    $00030409, // United States-Dvorak for left hand
+    $00040408, // Greek (319) Latin
+    $00040409, // United States-Dvorak for right hand
+    $00050408, // Greek Latin
+    $00050409  // US English Table for IBM Arabic 238_L
+  );
+
+// Tests if the reported base layout is known to be Latin script or not
+//
+// We work off a database of known layouts, then fallback to our cache. If not
+// found in the cache, or the cache has not yet been built, then we'll assume it
+// isn't a Latin script layout.
+//
+// The reason we do this is while we can determine algorithmically that a given
+// layout is Latin-script based, this involves a fairly costly load of the
+// layout DLL and enumeration of every key on the keyboard. We don't want to be
+// loading the layout DLL every time that the COM objects are instantiated; both
+// because of the cost and also to avoid any potential side-effects.
+//
+// If the user has chosen a layout that is not present in our known layouts
+// list, through the Base Keyboard dialog, we can safely assume the cache has
+// been built -- because the cache has to be populated in order to present the
+// list of available layouts in kmshell's BaseKeyboards classes.
+//
+class function TBaseKeyboard.IsLatinScriptLayout(Layout: DWORD): Boolean;
+var
+  i: DWord;
+  r: TRegistry;
+begin
+  // Check our known Latin script layouts first, because we may not yet
+  // have built our local cache
+  for i in KnownLatinScriptLayouts do
+    if i = Layout then
+      Exit(True);
+
+  // Otherwise, our Latin script keyboard cache in the registry tells us
+  // if this is a known Latin-script layout
+  r := TRegistry.Create;
+  try
+    r.RootKey := HKEY_LOCAL_MACHINE;
+    Result :=
+      r.OpenKeyReadOnly(SRegKey_LatinKeyboardCache_LM) and
+      r.ValueExists(IntToHex(Layout, 8)) and
+      (r.ReadString(SRegKey_LatinKeyboardCache_LM) = '1');
+  finally
+    r.Free;
+  end;
+end;
+
+//
+// Return only a valid Latin script base layout keyboard id
+//
+class function TBaseKeyboard.GetDefaultKeyboardID: DWORD;   // I4169
```

Comment:
If I'm making the right inferences, I believe this would be better renamed as `GetBaseLayoutKeyboardID` or something similar.

To me, "default keyboard" would always be the default system keyboard outside of Keyman, like Russian in the base issue's example.  "Base layout" would help make the connection with mnemonic functionality more clear.  Also... it's probably wise not to conflate the two separate roles if we can help it.